added all chats to show on teacher page

karlrez opened the pull request:
for #4

Initial draft of the UI for showing all chats on teacher page:

 
![image](https://user-images.githubusercontent.com/47090434/215287224-83f1dc8b-c4ba-4950-b6fb-0dbcfdd46b59.png)

@mssiegel what are your thoughts on this layout? 
- On mobile, the chats show right under the main "selected" chat which might cause some confusion. I think having a header or at least some spacing would be a good way to seperate them.
- On mobile, the user will scroll down and select one of the chats, do you think it would be a good idea to auto scroll the user back to the top of the page?
- Will you want the current selected chat to be removed from the bottom or should it stay in the list? I think we should keep it there and give it a different background color to indicate it is the selected chat.

zachlankton commented on src/components/pages/HomePage/ExampleChat.tsx:
```diff
@@ -30,7 +30,7 @@ export default function ExampleChat() {
   return (
     <>
-      <Chatbox chat={chat} />
+      <Chatbox chat={chat} showCopyButton={true} />
```
Nice addition and attention to detail

zachlankton commented on src/components/pages/TeachersPage/index.tsx:
```diff
@@ -20,6 +20,215 @@ interface StudentChat {
   startTime: string;
 }
 
+const EXAMPLE_CHATS = [
+  {
+    chatId: '8DmBxTmt0Q_iUbbxAAAn#Ua_hGB2pSD-6CGvdAAAP',
+    studentPair: [
+      {
+        realName: 'Student 1797',
```
nit: I would move this into its own file

karlrez replied:
I just removed it as I just intended to use it to display the layout, but if you find it helpful for dev purposes I can leave it as a seperate file

zachlankton replied:
ah gotcha, no, deleting is fine, just wasn't sure if this was supposed to be a part of the finished product

mssiegel replied:
@karlrez actually I do think we should keep these example chats for dev purposes. Can you put it as its own file? So other devs will realize this file exists, I think it should go under the TeachersPage folder.   You could leave a comment at the top of the file explaining the file isn't used in production but is there for dev purposes so that any developer can create a whole bunch of example chats to test out the UI.

zachlankton reviewed (commented):
I really like the look of the preview so far!

mssiegel commented on src/components/pages/TeachersPage/index.tsx:
```diff
@@ -156,6 +365,29 @@ export default function TeachersPage({ classroomName }: ClassroomProps) {
+        </Grid>
+        {/* change to studentChats.map() to view actual chats  */}
+        {EXAMPLE_CHATS.map((chat, i) => {
+          // alter chat here to only have last 5 messages
```
Love that the final version will show only the last several messages! Fantastic!

mssiegel commented on src/components/pages/TeachersPage/index.tsx:
```diff
@@ -156,6 +365,29 @@ export default function TeachersPage({ classroomName }: ClassroomProps) {
+      <Grid container spacing={2} mt={2} pb={2}>
+        <Grid item xs={12}>
+          <Typography variant='h5' sx={{ color: 'white' }}>
+            Click to view full chat
```
Love the idea of clicking to view the full chat.  Let's change text to to be more descriptive.  "Click any conversation below to display full chat above".

mssiegel reviewed (commented):
This layout looks fantastic!
Regarding your questions @karlrez:
1) I feel like a simple border line would be good separator. Maybe something like  "2px dashed silver" with width of maybe 30px. But choose whatever separator you feel is best. 
2) The school teachers run our learning platform on their laptops. So we can probably just ignore mobile version of Teachers page.   So you can put in autoscroll if its super simple, otherwise just ignore it.  
3) Love the idea of keeping the displayed chat in the list and using a differentiator to show which is the selected chat. I'm okay with whatever you choose as the differentiator, whether thats a different background color, a border etc.

karlrez commented:
Hey @mssiegel @zachlankton just made the updates from the comments. Let me know of any further edits that should be added in. 

Currently, I've run into a bug where the `scrollDown()` in Conversations is causing the entire page to scroll whenever a student sends a message, which would make the teachers page very difficult to use. I'll make another commit once I figure out a solution, but if you have any suggestions let me know!

zachlankton reviewed (approved):
Everything is looking great, once the scrolldown bug is fixed, this is approved

mssiegel commented on src/components/pages/HomePage/ExampleChat.tsx:
```diff
@@ -30,7 +30,7 @@ export default function ExampleChat() {
   return (
     <>
-      <Chatbox chat={chat} />
+      <Chatbox chat={chat} showCopyButton={true} selected={false} />
```
Hahaha.  You can delete this ExampleChat.tsx file entirely.  It was only for the Homepage and it stopped being used there once you made your PR to redesign the homepage.

mssiegel commented on src/components/pages/TeachersPage/Chatbox.tsx:
```diff
@@ -6,11 +6,14 @@ import chatboxCSS from './Chatbox.css';
 import CopyButton from '@components/shared/CopyButton';
 
-export default function Chatbox({ chat }) {
+export default function Chatbox({ chat, showCopyButton, selected }) {
```
I think we should rename `showCopyButton` to `isTheDisplayedChat`.  Its more self-explanatory. Right now, show copy button is the only difference. But in the future, there might be more differences between the main displayed chat and the other chats on the teachers page. This new variable name will remain relevant for those future differences.

zachlankton replied:
@mssiegel  do you mean rename `selected` to `isTheDisplayChat` ?

mssiegel replied:
Oh, good point. I'd forgotten that `selected` prop was there!  In that case, yes I think we should rename 'selected' to either `isTheDisplayedChat` or `isTheSelectedChat` because its more self-explanatory.    And I don't think we need `showCopyButton` because we could determine to show the copy button if its the selected chat.

karlrez replied:
Actually, I just realized I I will need two seperate variables to be able to apply the `scrollDown` only to the top displayed chat but not to the bottom one. 

I think if I rename them `isTheDisplayedChat` and `inAllStudentChatDisplay` I would be able to make this work and are more descriptive names.

mssiegel commented on src/components/pages/TeachersPage/index.tsx:
```diff
@@ -156,6 +156,42 @@ export default function TeachersPage({ classroomName }: ClassroomProps) {
           {showDisplayedChat()}
         </Grid>
       </Grid>
+      <Grid
```
I feel the multichat display should be its own component. Not sure what the best name for it is. Maybe `<AllStudentChatsDisplay>`?

mssiegel reviewed (changes requested):
@karlrez Regarding the `scrollDown` in Conversations issue.

The goal of the scroll down is to scoll down to the latest message. But I think the multi-chat display will show the latest messages anyways because it only shows the most recent ones.  Therefore, you can probably remove `scrollDown` from any chats in the multi-chat. Simply pass down `isTheDisplayedChat` to Conversations and only run scrollDown if its true.  

There might be another problem though of scrollDown in the displayed chat.  Lets say there are 15 student chats going on and the teacher is below looking at the multi chats. As new messages come to the displayedChat, the `scrollDown` might cause the view to scroll up to the main displayed chat.  That might be annoying for the teacher.      But it would only happen if there are many many student chats and if the teacher is scrolled down to the bottom.  Is that right?  Because if so, maybe we can leave this "bug" as is and only fix it later if actual teachers complain about it.

karlrez commented:
@mssiegel I've added the changes from the last comments. Only difference is regarding the `showCopyButton` prop which I changed to `inAllStudentsChatDisplay`. Let me know your thoughts, I can go back and update if needed.

mssiegel commented on src/components/pages/TeachersPage/exampleChats.js:
```diff
@@ -0,0 +1,213 @@
+ * Replace studentChats in ./index.tsx to test out UI with multiple chats during development
+ */
+
+export const EXAMPLE_CHATS = [
```
This should be a `.ts` file.  Then you can put `import { StudentChat } from '.';` at the top of this file and export the StudentChat interface form ./`index.tsx
By making it a typescript file it ensures that if anyone changes the StudentChat interface they'll get a warning to change this Example Chat also.

```suggestion
export const EXAMPLE_CHATS: StudentChat[] = [
```

mssiegel reviewed (changes requested):
Phenomenal. Thanks @karlrez. Its just about ready to go.   I left a few comments with minor changes.  

Also, I noticed a bug where if there's a lot of chats, the `scrollDown()` causes the teachers page to literally scroll down very low.  The fix is to change the `scrollDown()` function in `src/utils/classrooms.tsx` to include` ({ behavior: 'smooth', block: 'center' });`   The "block: center" makes it so it scrolls to the center of the last chat message.     Thanks in advance for adding this fix

karlrez commented on src/components/pages/TeachersPage/Conversation.tsx:
```diff
@@ -6,13 +6,14 @@ import { Box, Typography } from '@mui/material';
   useEffect(() => {
-    if (chat.chatId !== 'homepage sample chat') scrollDown(lastMessage);
-  }, [chat.chatId, chat.conversation]);
+    if (!inAllStudentsChatDisplay && chat.chatId !== 'homepage sample chat')
```
@mssiegel right that makes sense. Regarding the dependency array, would we want to keep the `chat.conversation` so that it scrolls on every new message? Or else it would only scroll when we switch the displayed chat.

mssiegel replied:
@karlrez thats right. Good catch!

karlrez commented:
@mssiegel just added those changes in. Only additional addition was changing the `inAllStudentsChatDisplay` prop to `inAllStudentChatsDisplay`, figured the spelling should match the component name

mssiegel reviewed (approved):
@karlrez Thank you! This looks fantastic!! Good to merge.
